crates/services/gas_price_service/src/v1/service.rs

```diff
@@ -136,63 +139,79 @@ where
     }
 
     fn validate_block_gas_capacity(
-        &self,
         block_gas_capacity: u64,
     ) -> anyhow::Result<NonZeroU64> {
         NonZeroU64::new(block_gas_capacity)
             .ok_or_else(|| anyhow!("Block gas capacity must be non-zero"))
     }
 
-    async fn set_metadata(&mut self) -> anyhow::Result<()> {
-        let metadata: UpdaterMetadata = self.algorithm_updater.clone().into();
-        self.metadata_storage
-            .set_metadata(&metadata)
-            .map_err(|err| anyhow!(err))
-    }
-
-    async fn handle_normal_block(
-        &mut self,
+    async fn handle_normal_block<'a>(
+        &'a mut self,
         height: u32,
         gas_used: u64,
         block_gas_capacity: u64,
         block_bytes: u64,
         block_fees: u64,
-    ) -> anyhow::Result<()> {
-        let capacity = self.validate_block_gas_capacity(block_gas_capacity)?;
+    ) -> anyhow::Result<()>
+    where
+        StorageTxProvider::Transaction<'a>:
+            UnrecordedBlocks + SetMetadataStorage + SetDaSequenceNumber,
+    {
+        let mut storage_tx = self.storage_tx_provider.begin_transaction()?;
+        let capacity = Self::validate_block_gas_capacity(block_gas_capacity)?;
+
+        for da_block_costs in self.da_block_costs_buffer.drain(..) {
+            tracing::debug!("Updating DA block costs: {:?}", da_block_costs);
+            self.algorithm_updater.update_da_record_data(
+                &da_block_costs.l2_blocks,
+                da_block_costs.blob_size_bytes,
+                da_block_costs.blob_cost_wei,
+                &mut storage_tx,
+            )?;
+            storage_tx
+                .set_sequence_number(
+                    &BlockHeight::from(height),
+                    da_block_costs.sequence_number,
+                )
+                .map_err(|err| anyhow!(err))?;
+        }
 
         self.algorithm_updater.update_l2_block_data(
             height,
             gas_used,
             capacity,
             block_bytes,
             block_fees as u128,
+            &mut storage_tx,
         )?;
 
-        self.set_metadata().await?;
-        Ok(())
-    }
-
-    async fn handle_da_block_costs(
-        &mut self,
-        da_block_costs: DaBlockCosts,
-    ) -> anyhow::Result<()> {
-        self.algorithm_updater.update_da_record_data(
-            &da_block_costs.l2_blocks,
-            da_block_costs.blob_size_bytes,
-            da_block_costs.blob_cost_wei,
-        )?;
-
-        self.set_metadata().await?;
+        let metadata = self.algorithm_updater.clone().into();
```

Comment:
AlgorithmUpdater have `algorithm()` method to get an algorithm. I find it strange to not have a `metadata()` but a `into()` to get metadata